relay-server/src/actors/processor.rs

```diff
@@ -1655,6 +1732,192 @@ impl Service for EnvelopeProcessorService {
+        let payload = serde_json::to_vec(&self.views).unwrap().into();
+        let quantities = self.quantities.clone();
+
+        self.views.clear();
+        self.quantities.clear();
+        self.remaining = self.max_size;
+
+        (payload, quantities)
+    }
+}
+
+/// An upstream request that submits metric buckets via HTTP.
+///
+/// This request is not awaited. It automatically tracks outcomes if the request is not received.
+#[derive(Debug)]
+struct SendMetricsRequest {
+    /// If the partition key is set, the request is marked with `X-Sentry-Relay-Shard`.
+    partition_key: Option<String>,
+    /// Serialized metric buckets without encoding applied, used for signing.
+    unencoded: Bytes,
+    /// Serialized metric buckets with the stated HTTP encoding applied.
+    encoded: Bytes,
+    /// Encoding (compression) of the payload.
+    http_encoding: HttpEncoding,
+    /// Information about the metric quantities in the payload for outcomes.
+    quantities: Vec<(Scoping, SourceQuantities)>,
+    /// Address of the outcome aggregator to send outcomes to on error.
+    outcome_aggregator: Addr<TrackOutcome>,
+}
+
+impl UpstreamRequest for SendMetricsRequest {
+    fn set_relay_id(&self) -> bool {
+        true
+    }
+
+    fn sign(&mut self) -> Option<Bytes> {
+        Some(self.unencoded.clone())
+    }
+
+    fn method(&self) -> reqwest::Method {
+        reqwest::Method::POST
+    }
+
+    fn path(&self) -> std::borrow::Cow<'_, str> {
+        "/api/0/relays/metrics/".into()
+    }
+
+    fn route(&self) -> &'static str {
+        "global_metrics"
+    }
+
+    fn build(&mut self, builder: &mut http::RequestBuilder) -> Result<(), http::HttpError> {
+        builder
+            .content_encoding(self.http_encoding)
+            .header_opt("X-Sentry-Relay-Shard", self.partition_key.as_ref())
+            .header(header::CONTENT_TYPE, b"application/json")
+            .body(self.encoded.clone());
+
+        Ok(())
+    }
+
+    fn respond(
+        self: Box<Self>,
+        result: Result<http::Response, UpstreamRequestError>,
+    ) -> Pin<Box<dyn Future<Output = ()> + Send + Sync>> {
+        Box::pin(async {
+            match result {
+                Ok(mut response) => {
+                    response.consume().await.ok();
+                }
+                // Request did not arrive, we are responsible for outcomes.
+                Err(error) if !error.is_received() => {
+                    for (scoping, quantities) in self.quantities {
+                        utils::reject_metrics(
+                            &self.outcome_aggregator,
+                            quantities,
+                            scoping,
+                            Outcome::Invalid(DiscardReason::Internal),
+                        );
+                    }
+                }
+                // Upstream is responsible to log outcomes.
+                Err(_received) => (),
+            }
+        })
+    }
+}
+
```

Comment:
There's a lot of metrics-specific methods here now, maybe we should move them to a submodule (as standalone functions) in a follow-up PR.